### ISSUE TITLE
${HOME}/.cache/Ganga/.used_versions

### ISSUE THREAD

**Issue by JacekHoleczek** [ganga-devs/ganga#1403]
It seems that ganga-7.1.9 unconditionally creates a "${HOME}/.cache/Ganga/.used_versions" file.
The old ganga-6.0.44 was creating the ".used_versions" file in the "gangadir" (I always set a ganga version specific "gangadir").
Could you, please, restore the old behaviour (so that ganga does not pollute "${HOME}/.cache").

**Comment by mesmith75** [comment]
There is a reason that we moved it to `~/.cache` - we wanted the record of used versions to not be dependent on the repository.

**Comment by JacekHoleczek** [comment]
I am running multiple different ganga versions. They can be running at the same time (each of them gets its own ".gangarc" and "gangadir", of course).

It seems that the current hardcoded location assumes that just one single ganga version is allowed.

Maybe you could introduce some kind of a "flag" in the ".gangarc"" (where to store the ".used_versions" file).

**Comment by mesmith75** [comment]
Why does this matter? The only thing the `.used_versions` is used for is to print the release notes and make a new .gangarc when the minor version increments.

**Comment by JacekHoleczek** [comment]
Well, it's exactly as you say ...

I copied the old private ganga-6.0.44 file:
```
cp -a /.../ganga-6.0.44/gangadir/.used_versions ${HOME}/.cache/Ganga/
```
and then I called the new ganga-7.1.9, which then immediately felt free to overwrite its own files:
```
INFO     reading config file /.../.gangarc
INFO     This is your first time running Ganga >=6.1.14
INFO     Now attempting to migrate your IPython history file
INFO     IPython history migrated successfully.
INFO     It appears that this is the first time you have run 7.1.9
INFO     Your ganga config file will be updated.
INFO     re-reading in old config for updating...
INFO     reading config file /.../.gangarc
INFO     Copied current config file to /.../.gangarc.00
INFO     Creating ganga config file /.../.gangarc
```
Note: the "config file /.../.gangarc" was ALREADY specific to ganga-7.1.9 so it SHOULD NOT have been touched / modified at all.

Could you, please, give me some patch file / hint which restores the old ".used_versions" behaviour?

Note: I really need to use different ganga versions and they should not interfere / know about one another.

**Comment by egede** [comment]
@JacekHoleczek Is this only a problem if running with version 6.0.44 of Ganga? If yes, then I am afraid that we will not use effort on supporting such a use case. Ganga 6.0.44 was replaced about 4 years ago!

**Comment by JacekHoleczek** [comment]
Sorry, it is NOT a problem of ganga-6.0.44 (which uses its private ".used_versions" file).
Anyhow, ganga-6.0.44 is my "production" version as it behaves relatively stable (with some patches, of course).

It is a problem in ganga-7.1.9 which uses the global "${HOME}/.cache/Ganga/.used_versions" file.
I do not have any other ganga version installed for the moment so I could only show you the problems that the new policy generates using this (very) old version.
I assume I will get the same problems in future, as soon as I install some new ganga version (which will then interfere with the currently installed ganga-7.1.9).

I really need to make them completely "separate" / "isolated".

Here's a patch for the ganga-7.1.9 release (Ganga-7-1-9):

```diff
--- GangaCore/Runtime/bootstrap.py.old	2019-01-13 10:14:00.000000000 +0000
+++ GangaCore/Runtime/bootstrap.py	2019-01-14 15:52:21.000000000 +0000
@@ -355,7 +355,8 @@
 
     @staticmethod
     def new_version():
-        versions_filename = os.path.expanduser('~/.cache/Ganga/.used_versions')
+        # versions_filename = os.path.expanduser('~/.cache/Ganga/.used_versions')
+        versions_filename = os.path.join(GangaCore.Utility.Config.getConfig('Configuration')['gangadir'], '.used_versions')
         if not os.path.exists(os.path.dirname(versions_filename)):
             os.makedirs(os.path.dirname(versions_filename))
         if not os.path.exists(versions_filename):
@@ -380,7 +381,8 @@
 
     @staticmethod
     def rollHistoryForward():
-        versions_filename = os.path.expanduser('~/.cache/Ganga/.used_versions')
+        # versions_filename = os.path.expanduser('~/.cache/Ganga/.used_versions')
+        versions_filename = os.path.join(GangaCore.Utility.Config.getConfig('Configuration')['gangadir'], '.used_versions')
         hasLoaded_newer = False
         with open(versions_filename, 'r') as versions_file:
             this_version = versions_file.read()
```

**Comment by mesmith75** [comment]
Why not use the latest ganga version?

**Comment by JacekHoleczek** [comment]
Up to now ganga-6.0.44 is the only version which works for me.
During the last several years, from time to time, I tried many "new releases" ... but up to now I always failed to get them running correctly (actually, I used ganga-6.3.1 for one of my "projects").
So, I simply need a stable "production" version which must not interfere with any games that I play with any other version.
Assuming that ganga-7.2.0 will be o.k. for me (after all recent fixes), I will then make it my "production" version for the next years ... but I will still need to make sure that it is completely "isolated" from the old ganga-6.0.44 (I will not delete it, it may happen that I will need to return to it) ... and it must be completely "isolated" from any new ganga versions which I will try to install and use in future.

And another point ... I try to use Ganga for several different "projects". Each "project" always gets its own private ".gangarc" and "gangadir". So, it may happen that some "project" will need some specific "old" Ganga version while another "project" will be fine with some "new" version. Again, I need to make sure that these "projects" (and their Ganga versions) are completely "isolated" from each other.

Well, I guess I really think that the "old" way of keeping the ".used_versions" in the "gangadir" was a good idea.
